processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```
@@ -0,0 +1,281 @@
+      initializeSlot(dimIndex);
+    }
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(valBuffer, recordOffset + aggregatorOffsets[i]);
+    }
+
+    return AggregateResult.ok();
+  }
+
+  private ByteBuffer checkAndGetKeyBuffer(KeySerde<KeyType> keySerde, KeyType key)
+  {
+    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
+    if (fromKey == null) {
+      return null;
+    }
+
+    if (fromKey.remaining() != keySerde.keySize()) {
+      throw new IAE(
+          "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+          fromKey.remaining(),
+          keySerde.keySize()
+      );
+    }
+    return fromKey;
+  }
+
+  private void initializeSlot(int dimIndex)
+  {
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    usedFlagBuffer.put(index, (byte) (usedFlagBuffer.get(index) | 1 << extraIndex));
+
+    final int recordOffset = dimIndex * recordSize;
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(valBuffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  private boolean isUsedSlot(int dimIndex)
+  {
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    final int usedByte = 1 << extraIndex;
+    return (usedFlagBuffer.get(index) & usedByte) == usedByte;
+  }
+
+  @Override
+  public void reset()
+  {
+    final int usedBufferCapacity = usedFlagBuffer.capacity();
+    for (int i = 0; i < usedBufferCapacity; i++) {
+      usedFlagBuffer.put(i, (byte) 0);
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Entry<KeyType>> iterator(boolean sorted)
+  {
+    // result is always natually sorted by keys
```

My bad. I think it would be better to throw UnsupportedOperationException if sorted is true for now. This can be removed when we use BufferArrayGrouper for merging partial aggregates.